crates/polars-ops/src/chunked_array/repeat_by.rs

```diff
@@ -111,6 +115,74 @@ fn repeat_by_binary(ca: &BinaryChunked, by: &IdxCa) -> PolarsResult<ListChunked>
     }
 }
 
+fn repeat_by_list(ca: &ListChunked, by: &IdxCa) -> PolarsResult<ListChunked> {
+    check_lengths(ca.len(), by.len())?;
+
+    match (ca.len(), by.len()) {
+        (left_len, right_len) if left_len == right_len => Ok(repeat_by_generic_inner(ca, by)),
+        (_, 1) => {
+            let by = new_by(by, ca.len());
+            repeat_by_list(ca, &by)
+        },
+        (1, _) => {
+            let new_array = ca.new_from_index(0, by.len());
+            repeat_by_list(&new_array, by)
+        },
+        // we have already checked the length
+        _ => unreachable!(),
+    }
+}
+
+#[cfg(feature = "dtype-struct")]
+fn repeat_by_struct(ca: &StructChunked, by: &IdxCa) -> PolarsResult<ListChunked> {
+    check_lengths(ca.len(), by.len())?;
+
+    match (ca.len(), by.len()) {
+        (left_len, right_len) if left_len == right_len => Ok(repeat_by_generic_inner(ca, by)),
+        (_, 1) => {
+            let by = new_by(by, ca.len());
+            repeat_by_struct(ca, &by)
+        },
+        (1, _) => {
+            let new_array = ca.new_from_index(0, by.len());
+            repeat_by_struct(&new_array, by)
+        },
+        // we have already checked the length
+        _ => unreachable!(),
+    }
+}
+
+fn repeat_by_generic_inner<T: PolarsDataType>(ca: &ChunkedArray<T>, by: &IdxCa) -> ListChunked {
+    arity::binary(ca, by, |arr, by| {
+        let output_length = by.iter().flatten().map(|x| *x as usize).sum();
+        let mut builder = make_builder(arr.dtype());
```

Comment:
Can you move the creation of `builder` outside of the inner loop? You'll also have to change `.freeze()` into `.freeze_reset()`. You'll need to compute the datatype to pass in yourself using `ca.dtype().to_arrow(CompatLevel::newest)`.